internal/protocol/connector.go

```diff
@@ -249,17 +252,33 @@ func Handshake(ctx context.Context, conn net.Conn, version uint64) (*Protocol, e
 		conn.SetDeadline(deadline)
 		defer conn.SetDeadline(time.Time{})
 	}
+	// Honor context cancellation.
+	var canceled int32 = 0
+	stop := utils.AfterFunc(ctx, func() {
+		if ctx.Err() == context.Canceled {
+			atomic.SwapInt32(&canceled, 1)
+			// Cancel read and writes by setting deadline in the past.
+			conn.SetDeadline(time.Unix(1, 0))
+		}
+	})
+	defer stop()
 
 	// Perform the protocol handshake.
+	// TODO: this should be a message and use same code for writing and reading
+	// to the network.
 	n, err := conn.Write(protocol)
 	if err != nil {
+		nerr, ok := errors.Cause(err).(net.Error)
```

Comment:
Naive Go question, do we know for sure that we can use `errors.Cause` here i.e. that the error from `conn.Write` (and in other places) conforms to the expected interface? Is this something stable that the standard library can't break?

Reply:
```
// Cause returns the underlying cause of the error, if possible.
// An error value has a cause if it implements the following
// interface:
//
//     type causer interface {
//            Cause() error
//     }
//
// If the error does not implement Cause, the original error will
// be returned. If the error is nil, nil will be returned without further
// investigation.
```

I would say it is safe to use because even if it is nil the type assertion should not work.